app/models/attachable_attachment.rb

```diff
@@ -0,0 +1,40 @@
+class AttachableAttachment < ActiveRecord::Base
+  belongs_to :attachable, polymorphic: true, inverse_of: :attachable_attachments
+  belongs_to :attachment, inverse_of: nil
+
+  # Get the name from the file and then further build or find an attachment based on file's SHA256
+  # hash.
+  #
+  # @param [File|ActionDispatch::Http::UploadedFile] The uploaded file.
+  def file=(file)
+    self.name = filename(file)
+    self.attachment = find_or_initialize_attachment_by(file)
+  end
+
+  private
+
+  # Calculates the SHA256 hash and return the existing attachment which has the same hash.
+  # Otherwise a new attachment will be built.
+  #
+  # @param [File|ActionDispatch::Http::UploadedFile] The uploaded file.
+  # @return [Attachment] The attachment which contains the file.
+  def find_or_initialize_attachment_by(file)
+    # Get the actual file by #tempfile if the file is an `ActionDispatch::Http::UploadedFile`.
+    digest = Digest::SHA256.file(file.try(:tempfile) || file).hexdigest
+    Attachment.find_by(name: digest) || build_attachment(name: digest, file_upload: file)
```

Comment:
the attachment hash name has a unique index, so in that case the later one won't be saved.